File: antlr/parser/ProgramDtoBuilder.js
const PseudocodeVisitor = require('./PseudocodeVisitor')

function ProgramDtoBuilder() {
	PseudocodeVisitor.PseudocodeVisitor.call(this)
	return this
}

ProgramDtoBuilder.prototype = Object.create(PseudocodeVisitor.PseudocodeVisitor.prototype)
ProgramDtoBuilder.prototype.constructor = ProgramDtoBuilder


ProgramDtoBuilder.prototype.visitProgram = function(ctx) {
  // `ctx.statements()` fails when `ctx.children` is `null`,
  // which does happen for an empty program.
  // This is a bug. `children` should probably be an empty array.
  // So do not replace `ctx.children` with `ctx.statements()`.
  // console.log(ctx.children)
  return { statements: ctx.children !== null ? this.visit(ctx.statements()) : [] }
}


ProgramDtoBuilder.prototype.visitStatements = function(ctx) {
  // console.log('Statements')
  return ctx.statement().map(sctx => {
    const statement = this.visit(sctx)

    statement.textRange = {
      start: {
        row: sctx.start.line,
        col: sctx.start.column
      },
      end: {
        row: sctx.stop.line,
        col: sctx.stop.column 
             + sctx.stop.text.length
      }
    }

    return statement
  })
}


ProgramDtoBuilder.prototype.visitAbort = function(ctx) {
  // console.log('Abort')
  return { type: 'abort' }
}


ProgramDtoBuilder.prototype.visitSkip = function(ctx) {
  // console.log('Skip')
  return { type: 'skip' }
}


ProgramDtoBuilder.prototype.visitAssignment_statement = function(ctx) {
  // console.log('Assignment_statement')
  const assignment = this.visit(ctx.assignment())
  assignment.type = 'assign'
  return assignment
}


ProgramDtoBuilder.prototype.visitScalar_assignment = function(ctx) {
  // console.log('Scalar_assignment')
  return {
    lvalues: [this.visit(ctx.variable())],
    rvalues: [this.visit(ctx.int_expr())]
  }
}


ProgramDtoBuilder.prototype.visitVector_assignment = function(ctx) {
  // console.log('Vector_assignment')
  const assignment = this.visit(ctx.assignment())
  assignment.lvalues.unshift(this.visit(ctx.variable()))
  assignment.rvalues.push(this.visit(ctx.int_expr()))
  return assignment
}


ProgramDtoBuilder.prototype.visitIf_statement = function(ctx) {
  // console.log('If_statement')
  const guardedCommands = this.visit(ctx.guarded_commands())
  guardedCommands.type = 'if'
  return guardedCommands
}


ProgramDtoBuilder.prototype.visitDo_statement = function(ctx) {
  // console.log('Do_statement')
  const guardedCommands = this.visit(ctx.guarded_commands())
  guardedCommands.type = 'do'
  return guardedCommands
}


ProgramDtoBuilder.prototype.visitGuarded_commands = function(ctx) {
  // console.log('Guarded_commands')
  return {
    guards: ctx.guarded_command().map(gc => this.visit(gc.bool_expr())),
    commands: ctx.guarded_command().map(gc => this.visit(gc.statements()))
  }
}


ProgramDtoBuilder.prototype.visitInt_const_expr = function(ctx) {
  // console.log('Int_const_expr')
  return {
    type: 'const',
    negated: ctx.MINUS() != null,
    const: parseInt(ctx.INT().getText())
  }
}


ProgramDtoBuilder.prototype.visitVariable_expr = function(ctx) {
  // console.log('Variable_expr')
  return {
    type: 'var',
    negated: ctx.MINUS() != null,
    var: this.visit(ctx.variable())
  }
}


ProgramDtoBuilder.prototype.visitParet_int_expr = function(ctx) {
  // console.log('Paret_int_expr')
  return {
    type: 'parets',
    negated: ctx.MINUS() != null,
    inner: this.visit(ctx.int_expr())
  }
}


ProgramDtoBuilder.prototype.visitMult_expr = function(ctx) {
  // console.log('Mult_expr')
  return {
    type: 'mult',
    left: this.visit(ctx.int_expr(0)),
    right: this.visit(ctx.int_expr(1))    
  }
}


ProgramDtoBuilder.prototype.visitAdd_expr = function(ctx) {
  // console.log('Add_expr')
  return {
    type: ctx.PLUS() ? 'plus' : 'minus',
    left: this.visit(ctx.int_expr(0)),
    right: this.visit(ctx.int_expr(1))    
  }
}


ProgramDtoBuilder.prototype.visitBool_const_expr = function(ctx) {
  // console.log('Bool_const_expr')
  return {
    type: 'const',
    negated: ctx.NEGATION() != null,
    const: ctx.TRUE() != null
  }
}


ProgramDtoBuilder.prototype.visitParet_bool_expr = function(ctx) {
  // console.log('Paret_bool_expr')
  return {
    type: 'parets',
    negated: ctx.NEGATION() != null,
    inner: this.visit(ctx.bool_expr())
  }
}


ProgramDtoBuilder.prototype.visitAnd_expr = function(ctx) {
  // console.log('And_expr')
  return {
    type: 'and',
    leftBool: this.visit(ctx.bool_expr(0)),
    rightBool: this.visit(ctx.bool_expr(1))
  }
}


ProgramDtoBuilder.prototype.visitOr_expr = function(ctx) {
  // console.log('Or_expr')
  return {
    type: 'or',
    leftBool: this.visit(ctx.bool_expr(0)),
    rightBool: this.visit(ctx.bool_expr(1))
  }
}


ProgramDtoBuilder.prototype.visitComparison_expr = function(ctx) {
  // console.log('Comparison_expr')
  return {
    type: 'comp',
    leftInt: this.visit(ctx.int_expr(0)),
    rightInt: this.visit(ctx.int_expr(1)),
    comp: this.visit(ctx.comparison_op())
  }
}

ProgramDtoBuilder.prototype.visitLt  = ctx => '<'
ProgramDtoBuilder.prototype.visitGt  = ctx => '>'
ProgramDtoBuilder.prototype.visitLeq = ctx => '<='
ProgramDtoBuilder.prototype.visitGeq = ctx => '>='
ProgramDtoBuilder.prototype.visitEq  = ctx => '='
ProgramDtoBuilder.prototype.visitNeq = ctx => '<>'


ProgramDtoBuilder.prototype.visitVariable = function(ctx) {
  // console.log('Variable')
  return {
    name: ctx.NAME().getText(),
    selectors: ctx.selectors()
               ? this.visit(ctx.selectors())
               : []
  }
}


ProgramDtoBuilder.prototype.visitSelectors = function(ctx) {
  // console.log('Selectors')
  return ctx.selector()
    ? ctx.selector().map(sctx => this.visit(sctx))
    : []
}


ProgramDtoBuilder.prototype.visitSelector = function(ctx) {
  // console.log('Selector')
  return this.visit(ctx.int_expr())
}


module.exports = ProgramDtoBuilder

File: antlr/old-java-code/ProgramDtoBuilder.java
import java.util.*;
import org.antlr.v4.runtime.*;

public class ProgramDtoBuilder extends PseudocodeBaseVisitor<Object> {

    public Program buildProgramDto(ParserRuleContext programContext) {
        if (programContext instanceof PseudocodeParser.ProgramContext) {
            return visitProgram((PseudocodeParser.ProgramContext)programContext);
        }
        return null;
    }

    @Override
    public Program visitProgram(PseudocodeParser.ProgramContext ctx) {
        Program program = new Program();

        if (ctx.statements() != null) {
            program.statements = (List<Statement>)visit(ctx.statements());
        }

        return program;
    }
    
    @Override
    public List<Statement> visitStatements(PseudocodeParser.StatementsContext ctx) {
        List<Statement> statements = new ArrayList<>();

        for (PseudocodeParser.StatementContext statementContext : ctx.statement()) {
            Statement statement = (Statement)visit(statementContext);
            statement.textRange = getTextRange(statementContext);
            statements.add(statement);
        }

        return statements;
    }

    private TextRange getTextRange(ParserRuleContext ctx) {
        Token start = ctx.getStart();
        int startRow = start.getLine();
        int startCol = start.getCharPositionInLine();

        Token end = ctx.getStop();
        int endRow = end.getLine();
        int endCol = end.getCharPositionInLine() + end.getText().length();

        TextPosition startPosition = new TextPosition(startRow, startCol);
        TextPosition endPosition = new TextPosition(endRow, endCol);
        
        return new TextRange(startPosition, endPosition);
    }

    @Override
    public AbortStatement visitAbort(PseudocodeParser.AbortContext ctx) {
        return new AbortStatement();
    }

    @Override
    public SkipStatement visitSkip(PseudocodeParser.SkipContext ctx) {
        return new SkipStatement();
    }

    @Override
    public AssignmentStatement visitAssignment_statement(PseudocodeParser.Assignment_statementContext ctx) {
        return (AssignmentStatement)visit(ctx.assignment());
    }

    @Override
    public AssignmentStatement visitScalar_assignment(PseudocodeParser.Scalar_assignmentContext ctx) {
        AssignmentStatement assignment = new AssignmentStatement();
        assignment.lvalues.add((Variable)visit(ctx.var()));
        assignment.rvalues.add((IntegerExpression)visit(ctx.int_expr()));
        
        return assignment;
    }

    @Override
    public AssignmentStatement visitVector_assignment(PseudocodeParser.Vector_assignmentContext ctx) {
        AssignmentStatement assignment = (AssignmentStatement)visit(ctx.assignment());
        assignment.lvalues.add(0, (Variable)visit(ctx.var()));
        assignment.rvalues.add((IntegerExpression)visit(ctx.int_expr()));

        return assignment;
    }

    @Override
    public IfStatement visitIf_statement(PseudocodeParser.If_statementContext ctx) {
        IfStatement ifStatement = new IfStatement();
        addGuardedCommandsToContainer(ifStatement, ctx.guarded_commands());
        return ifStatement;
    }

    @Override
    public DoStatement visitDo_statement(PseudocodeParser.Do_statementContext ctx) {
        DoStatement doStatement = new DoStatement();
        addGuardedCommandsToContainer(doStatement, ctx.guarded_commands());
        return doStatement;
    }

    private void addGuardedCommandsToContainer(
        GuardedCommandsContainerStatement container,
        PseudocodeParser.Guarded_commandsContext ctx)
    {
        for (PseudocodeParser.Guarded_commandContext guardedCommand : ctx.guarded_command()) {
            container.guards.add((BooleanExpression)visit(guardedCommand.bool_expr()));
            container.commands.add((List<Statement>)visit(guardedCommand.statements()));
        }
    }

    @Override
    public IntegerParethesesExpression visitParet_int_expr(PseudocodeParser.Paret_int_exprContext ctx) {
        IntegerParethesesExpression expression = new IntegerParethesesExpression();
        expression.isNegated = ctx.MINUS() != null;
        expression.innerExpression = (IntegerExpression)visit(ctx.int_expr());

        return expression;
    }

    @Override
    public VariableExpression visitVariable_expr(PseudocodeParser.Variable_exprContext ctx) {
        VariableExpression expression = new VariableExpression();
        expression.isNegated = ctx.MINUS() != null;
        expression.variable = (Variable)visit(ctx.var());

        return expression;
    }

    @Override
    public IntegerConstantExpression visitInt_const_expr(PseudocodeParser.Int_const_exprContext ctx) {
        IntegerConstantExpression expression = new IntegerConstantExpression();
        expression.isNegated = ctx.MINUS() != null;
        expression.constant = ctx.INT().getText();

        return expression;
    }

    @Override
    public IntegerBinatyOperatorExpression visitAdd_expr(PseudocodeParser.Add_exprContext ctx) {
        IntegerBinatyOperatorExpression expression = new IntegerBinatyOperatorExpression();
        expression.leftHandSideExpression = (IntegerExpression)visit(ctx.int_expr(0));
        expression.rightHandSideExpression = (IntegerExpression)visit(ctx.int_expr(1));
        expression.operator = ctx.PLUS() != null ? IntegerOperators.PLUS : IntegerOperators.MINUS;
        
        return expression;
    }

    @Override
    public IntegerBinatyOperatorExpression visitMult_expr(PseudocodeParser.Mult_exprContext ctx) {
        IntegerBinatyOperatorExpression expression = new IntegerBinatyOperatorExpression();
        expression.leftHandSideExpression = (IntegerExpression)visit(ctx.int_expr(0));
        expression.rightHandSideExpression = (IntegerExpression)visit(ctx.int_expr(1));
        expression.operator = IntegerOperators.MULTIPLY;
        
        return expression;
    }

    @Override
    public BooleanConstantExpression visitBool_const_expr(PseudocodeParser.Bool_const_exprContext ctx) {
        BooleanConstantExpression expression = new BooleanConstantExpression();
        expression.isNegated = ctx.NEGATION() != null;
        expression.constant = ctx.TRUE() != null ? true : false;

        return expression;
    }

    @Override
    public BooleanBinaryOperatorExpression visitAnd_expr(PseudocodeParser.And_exprContext ctx) {
        BooleanBinaryOperatorExpression expression = new BooleanBinaryOperatorExpression();
        expression.leftHandSideExpression = (BooleanExpression)visit(ctx.bool_expr(0));
        expression.rightHandSideExpression = (BooleanExpression)visit(ctx.bool_expr(1));
        expression.operator = BooleanOperators.AND;

        return expression;
    }

    @Override
    public BooleanParethesesExpression visitParet_bool_expr(PseudocodeParser.Paret_bool_exprContext ctx) {
        BooleanParethesesExpression expression = new BooleanParethesesExpression();
        expression.isNegated = ctx.NEGATION() != null;
        expression.innerExpression = (BooleanExpression)visit(ctx.bool_expr());

        return expression;
    }

    @Override
    public BooleanBinaryOperatorExpression visitOr_expr(PseudocodeParser.Or_exprContext ctx) {
        BooleanBinaryOperatorExpression expression = new BooleanBinaryOperatorExpression();
        expression.leftHandSideExpression = (BooleanExpression)visit(ctx.bool_expr(0));
        expression.rightHandSideExpression = (BooleanExpression)visit(ctx.bool_expr(1));
        expression.operator = BooleanOperators.OR;

        return expression;
    }

    @Override
    public ComparisonExpression visitComparison_expr(PseudocodeParser.Comparison_exprContext ctx) {
        ComparisonExpression expression = new ComparisonExpression();
        expression.leftHandSideExpression = (IntegerExpression)visit(ctx.int_expr(0));
        expression.rightHandSideExpression = (IntegerExpression)visit(ctx.int_expr(1));
        expression.operator = (ComparisonOperators)visit(ctx.comparison_op());
        
        return expression;
    }

    @Override
    public ComparisonOperators visitLt(PseudocodeParser.LtContext ctx) {
        return ComparisonOperators.LESS;
    }

    @Override
    public ComparisonOperators visitGt(PseudocodeParser.GtContext ctx) {
        return ComparisonOperators.GREATER;
    }

    @Override
    public ComparisonOperators visitLeq(PseudocodeParser.LeqContext ctx) {
        return ComparisonOperators.LESS_OR_EQ;
    }

    @Override
    public ComparisonOperators visitGeq(PseudocodeParser.GeqContext ctx) {
        return ComparisonOperators.GREATER_OR_EQ;
    }

    @Override
    public ComparisonOperators visitEq(PseudocodeParser.EqContext ctx) {
        return ComparisonOperators.EQ;
    }

    @Override
    public ComparisonOperators visitNeq(PseudocodeParser.NeqContext ctx) {
        return ComparisonOperators.NOT_EQ;
    }

    @Override
    public Variable visitVar(PseudocodeParser.VarContext ctx) {
        Variable variable = new Variable();
        variable.name = ctx.NAME().getText();

        if (ctx.selectors() != null) {        
            variable.selectors = (List<IntegerExpression>)visit(ctx.selectors());
        }
        
        return variable;
    }

    @Override
    public List<IntegerExpression> visitSelectors(PseudocodeParser.SelectorsContext ctx) {
        List<IntegerExpression> selectors = new ArrayList<>();

        for (PseudocodeParser.SelectorContext selectorCtx : ctx.selector()) {
            selectors.add((IntegerExpression)visit(selectorCtx));
        }

        return selectors;
    }

    @Override
    public IntegerExpression visitSelector(PseudocodeParser.SelectorContext ctx) {
        return (IntegerExpression)visit(ctx.int_expr());
    }
}
